File: app-android/vela-wallet/app/src/main/java/app/getvela/wallet/feature/onboarding/core/HybridCableScanner.kt
package app.getvela.wallet.feature.onboarding.core

import android.annotation.SuppressLint
import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGattCallback
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothManager
import android.bluetooth.BluetoothProfile
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanFilter
import android.bluetooth.le.ScanResult
import android.bluetooth.le.ScanSettings
import android.content.Context
import android.os.Handler
import android.os.Looper
import android.os.ParcelUuid
import app.getvela.wallet.core.diagnostics.VelaLog
import kotlinx.coroutines.suspendCancellableCoroutine
import uniffi.vela_core_uniffi.CableAdvert
import uniffi.vela_core_uniffi.cableTryDecryptAdvert
import java.util.UUID
import kotlin.coroutines.resume

/**
 * Finds the phone that scanned OUR QR, by its BLE proximity advert — a port of
 * the founder's proven demo (`apppasskeysdemo/.../HybridBleClient.findResponder`),
 * with the crypto moved into vela-core: every candidate goes through
 * `cableTryDecryptAdvert`, which answers with the decrypted EID and the L2CAP
 * PSM in one step.
 *
 * Two discovery shapes, because the platforms advertise differently:
 *
 *  * **service data** under 0xFFF9/0xFDE2 — Android authenticators (GMS,
 *    securitykeys) broadcast the sealed EID directly;
 *  * **bare service UUID** — iOS cannot broadcast service data, so an iOS
 *    authenticator advertises just 0xFFF9 and publishes the EID + PSM on GATT
 *    characteristic 0xFFFA, read after a short connect.
 */
class HybridCableScanner(context: Context) {
    private val appContext = context.applicationContext
    private val adapter: BluetoothAdapter? =
        (context.getSystemService(Context.BLUETOOTH_SERVICE) as? BluetoothManager)?.adapter

    fun bluetoothReady(): Boolean = adapter?.isEnabled == true

    /** The matched phone: where to connect, and what its advert said. */
    data class AdvertHit(
        val device: BluetoothDevice,
        val advert: CableAdvert,
    )

    @SuppressLint("MissingPermission")
    suspend fun findResponder(qrSecret: ByteArray, timeoutMs: Long): AdvertHit? =
        suspendCancellableCoroutine { cont ->
            val scanner = adapter?.bluetoothLeScanner
            if (scanner == null) {
                VelaLog.event("cable.scan", "bluetooth unavailable")
                if (cont.isActive) cont.resume(null)
                return@suspendCancellableCoroutine
            }
            val handler = Handler(Looper.getMainLooper())
            var finished = false
            var gattStarted = false
            lateinit var cb: ScanCallback
            fun finish(result: AdvertHit?) {
                if (finished) return
                finished = true
                handler.removeCallbacksAndMessages(null)
                runCatching { scanner.stopScan(cb) }
                if (cont.isActive) cont.resume(result)
            }
            cb = object : ScanCallback() {
                override fun onScanResult(callbackType: Int, result: ScanResult) {
                    val rec = result.scanRecord ?: return
                    for (uuid in SERVICE_UUIDS) {
                        val sd = rec.getServiceData(uuid) ?: continue
                        if (sd.size < 20) continue
                        // The WHOLE payload: bytes past 20 are the CTAP 2.3 BLE
                        // suffix the core reads the PSM from.
                        val advert = cableTryDecryptAdvert(qrSecret, sd) ?: continue
                        VelaLog.event(
                            "cable.scan",
                            "matched this QR",
                            "rssi" to result.rssi,
                            "psm" to (advert.psm ?: "none"),
                        )
                        finish(AdvertHit(result.device, advert))
                        return
                    }
                    // No service data but our service UUID: an iOS authenticator.
                    // Read the EID/PSM from its GATT proximity characteristic.
                    if (!gattStarted && rec.serviceUuids?.any { it in SERVICE_UUIDS } == true) {
                        gattStarted = true
                        runCatching { scanner.stopScan(cb) }
                        VelaLog.event("cable.scan", "bare 0xFFF9 UUID (iOS authenticator?) — reading GATT 0xFFFA")
                        readAdvertViaGatt(result.device, qrSecret) { hit -> finish(hit) }
                    }
                }

                override fun onScanFailed(errorCode: Int) {
                    VelaLog.event("cable.scan", "scan failed", "code" to errorCode)
                    finish(null)
                }
            }
            // MIUI silently throttles UNFILTERED scans (results never arrive);
            // one filter per shape makes them get delivered. Device-found in the
            // demo era on a Xiaomi, kept because that lesson was expensive.
            val settings = ScanSettings.Builder()
                .setScanMode(ScanSettings.SCAN_MODE_LOW_LATENCY)
                .build()
            val filters = SERVICE_UUIDS.map {
                ScanFilter.Builder().setServiceData(it, ByteArray(0)).build()
            } + SERVICE_UUIDS.map {
                ScanFilter.Builder().setServiceUuid(it).build()
            }
            VelaLog.event("cable.scan", "scanning 0xFFF9/0xFDE2")
            scanner.startScan(filters, settings, cb)
            handler.postDelayed({ finish(null) }, timeoutMs)
            cont.invokeOnCancellation { finish(null) }
        }

    /**
     * iOS fallback: the EID (+ suffix) lives on GATT characteristic 0xFFFA.
     * Calls [onDone] exactly once, with a matching hit or null.
     */
    @SuppressLint("MissingPermission")
    private fun readAdvertViaGatt(
        device: BluetoothDevice,
        qrSecret: ByteArray,
        onDone: (AdvertHit?) -> Unit,
    ) {
        var settled = false
        fun settle(hit: AdvertHit?, gatt: BluetoothGatt?) {
            if (settled) return
            settled = true
            runCatching { gatt?.disconnect() }
            runCatching { gatt?.close() }
            onDone(hit)
        }
        val cb = object : BluetoothGattCallback() {
            override fun onConnectionStateChange(g: BluetoothGatt, status: Int, newState: Int) {
                if (newState == BluetoothProfile.STATE_CONNECTED) {
                    // Android caches the GATT database across connections; a
                    // service the peer only just published would look absent
                    // from a stale cache. The hidden-but-ubiquitous refresh().
                    runCatching { g.javaClass.getMethod("refresh").invoke(g) }
                    if (!g.requestMtu(185)) g.discoverServices()
                } else if (newState == BluetoothProfile.STATE_DISCONNECTED) {
                    settle(null, g)
                }
            }

            override fun onMtuChanged(g: BluetoothGatt, mtu: Int, status: Int) {
                g.discoverServices()
            }

            override fun onServicesDiscovered(g: BluetoothGatt, status: Int) {
                val ch = SERVICE_UUIDS.firstNotNullOfOrNull {
                    g.getService(it.uuid)?.getCharacteristic(PROX_CHAR_UUID)
                }
                if (ch == null || !g.readCharacteristic(ch)) settle(null, g)
            }

            @Suppress("DEPRECATION")
            override fun onCharacteristicRead(g: BluetoothGatt, ch: BluetoothGattCharacteristic, status: Int) {
                if (status != BluetoothGatt.GATT_SUCCESS) {
                    settle(null, g)
                    return
                }
                val bytes = ch.value ?: ByteArray(0)
                if (bytes.size < 20) {
                    settle(null, g)
                    return
                }
                val advert = cableTryDecryptAdvert(qrSecret, bytes)
                if (advert == null) {
                    settle(null, g)
                    return
                }
                VelaLog.event("cable.scan", "GATT matched this QR", "psm" to (advert.psm ?: "none"))
                settle(AdvertHit(device, advert), g)
            }
        }
        device.connectGatt(appContext, false, cb, BluetoothDevice.TRANSPORT_LE)
    }

    private companion object {
        val SERVICE_UUIDS = listOf(
            ParcelUuid(UUID.fromString("0000fff9-0000-1000-8000-00805f9b34fb")),
            ParcelUuid(UUID.fromString("0000fde2-0000-1000-8000-00805f9b34fb")),
        )
        val PROX_CHAR_UUID: UUID = UUID.fromString("0000fffa-0000-1000-8000-00805f9b34fb")
    }
}

File: app-ios/VelaWallet/VelaWallet/Features/Onboarding/Core/HybridCableScanner.swift
//
//  HybridCableScanner.swift
//  VelaWallet
//
//  Finds the phone that scanned OUR QR, by its BLE proximity advert — a port of
//  the founder's proven demo (apppasskeysdemo-ios/HybridBleClient.swift), with
//  the crypto moved into vela-core: every candidate goes through
//  cableTryDecryptAdvert, which answers with the decrypted EID and the L2CAP
//  PSM in one step.
//
//  iOS constraint (vs Android's connection-less createInsecureL2capChannel):
//  openL2CAPChannel needs a prior central.connect(peripheral), which requires
//  the responder to advertise CONNECTABLY — Android responders (securitykeys,
//  the demo) do.
//

import Foundation
import CoreBluetooth
import VelaCore

@MainActor
final class HybridCableScanner: NSObject, CBCentralManagerDelegate, CBPeripheralDelegate {

    private static let serviceUUIDs = [CBUUID(string: "FFF9"), CBUUID(string: "FDE2")]

    /// The matched phone: where to connect, and what its advert said.
    struct AdvertHit {
        let peripheral: CBPeripheral
        let advert: CableAdvert
    }

    private var central: CBCentralManager!
    private var qrSecret = Data()

    private var scanCont: CheckedContinuation<AdvertHit?, Never>?
    private var scanWatchdog: Task<Void, Never>?

    private var connectCont: CheckedContinuation<Void, Error>?
    private var openCont: CheckedContinuation<CBL2CAPChannel, Error>?
    private var connectWatchdog: Task<Void, Never>?
    private var connectingPeripheral: CBPeripheral?

    override init() {
        super.init()
        central = CBCentralManager(delegate: self, queue: nil)
    }

    var isBluetoothReady: Bool { central.state == .poweredOn }

    // MARK: scan

    func findResponder(qrSecret: Data, timeoutMs: Int) async -> AdvertHit? {
        self.qrSecret = qrSecret
        return await withCheckedContinuation { (cont: CheckedContinuation<AdvertHit?, Never>) in
            self.scanCont = cont
            self.startScanIfReady()
            self.scanWatchdog = Task { @MainActor [weak self] in
                try? await Task.sleep(nanoseconds: UInt64(timeoutMs) * 1_000_000)
                self?.finishScan(nil)
            }
        }
    }

    private func startScanIfReady() {
        // Wait for didUpdateState when the radio is still warming up.
        guard central.state == .poweredOn else {
            print("[vela-cable] bluetooth not ready (state=\(central.state.rawValue)); waiting")
            return
        }
        print("[vela-cable] scan started (no service filter, trial-decrypting service data)")
        // `withServices: nil`, NOT the FIDO UUIDs: the caBLE proximity advert
        // carries them inside Service Data (AD type 0x16), and CoreBluetooth's
        // service filter matches only the advertised service-UUID list — it
        // would drop a service-data-only responder. Every advert is delivered
        // and the per-advert trial-decrypt is the real filter.
        central.scanForPeripherals(
            withServices: nil,
            options: [CBCentralManagerScanOptionAllowDuplicatesKey: true]
        )
    }

    private func finishScan(_ hit: AdvertHit?) {
        guard let cont = scanCont else { return }
        scanCont = nil
        scanWatchdog?.cancel(); scanWatchdog = nil
        if central.state == .poweredOn { central.stopScan() }
        cont.resume(returning: hit)
    }

    nonisolated func centralManagerDidUpdateState(_ c: CBCentralManager) {
        MainActor.assumeIsolated {
            print("[vela-cable] bluetooth state -> \(c.state.rawValue)")
            switch c.state {
            case .poweredOn: if scanCont != nil { startScanIfReady() }
            case .unauthorized, .poweredOff:
                print("[vela-cable] bluetooth unavailable/denied — scan over")
                finishScan(nil)
            default: break
            }
        }
    }

    nonisolated func centralManager(
        _ c: CBCentralManager,
        didDiscover p: CBPeripheral,
        advertisementData: [String: Any],
        rssi RSSI: NSNumber
    ) {
        MainActor.assumeIsolated {
            guard let sdMap = advertisementData[CBAdvertisementDataServiceDataKey] as? [CBUUID: Data] else { return }
            for uuid in Self.serviceUUIDs {
                guard let sd = sdMap[uuid], sd.count >= 20 else { continue }
                // The WHOLE payload: bytes past 20 are the CTAP 2.3 BLE suffix
                // the core reads the PSM from.
                guard let advert = cableTryDecryptAdvert(qrSecret: qrSecret, candidate: sd) else {
                    print("[vela-cable] caBLE advert for a DIFFERENT QR (\(sd.count) bytes)")
                    continue
                }
                print("[vela-cable] matched this QR; PSM=\(advert.psm.map(String.init) ?? "none") rssi=\(RSSI)")
                connectingPeripheral = p
                finishScan(AdvertHit(peripheral: p, advert: advert))
                return
            }
        }
    }

    // MARK: connect + open the CTAP 2.3 BLE channel

    /// GATT-connect the matched peripheral and open the L2CAP CoC at `psm`.
    func openL2cap(_ hit: AdvertHit, psm: UInt16, timeoutMs: Int) async throws -> L2capCableConn {
        let p = hit.peripheral
        p.delegate = self
        connectingPeripheral = p

        // Connect with a watchdog — iOS never times out connect() by itself.
        try await withCheckedThrowingContinuation { (cont: CheckedContinuation<Void, Error>) in
            self.connectCont = cont
            self.central.connect(p, options: nil)
            self.connectWatchdog = Task { @MainActor [weak self] in
                try? await Task.sleep(nanoseconds: UInt64(timeoutMs) * 1_000_000)
                guard let self, let c = self.connectCont else { return }
                self.connectCont = nil
                self.central.cancelPeripheralConnection(p)
                c.resume(throwing: CableConnError.timeout)
            }
        }
        connectWatchdog?.cancel(); connectWatchdog = nil
        print("[vela-cable] GATT connected; opening L2CAP CoC (PSM \(psm))")

        let channel = try await withCheckedThrowingContinuation { (cont: CheckedContinuation<CBL2CAPChannel, Error>) in
            self.openCont = cont
            p.openL2CAPChannel(CBL2CAPPSM(psm))
        }
        print("[vela-cable] L2CAP channel open")
        return L2capCableConn(channel, peripheral: p)
    }

    nonisolated func centralManager(_ c: CBCentralManager, didConnect p: CBPeripheral) {
        MainActor.assumeIsolated {
            guard let cont = connectCont else { return }
            connectCont = nil
            cont.resume(returning: ())
        }
    }

    nonisolated func centralManager(_ c: CBCentralManager, didFailToConnect p: CBPeripheral, error: Error?) {
        MainActor.assumeIsolated {
            guard let cont = connectCont else { return }
            connectCont = nil
            cont.resume(throwing: CableConnError.connectFailed(error?.localizedDescription ?? "didFailToConnect"))
        }
    }

    nonisolated func centralManager(_ c: CBCentralManager, didDisconnectPeripheral p: CBPeripheral, error: Error?) {
        MainActor.assumeIsolated {
            if let cont = connectCont { connectCont = nil; cont.resume(throwing: CableConnError.closed) }
            if let cont = openCont { openCont = nil; cont.resume(throwing: CableConnError.closed) }
        }
    }

    nonisolated func peripheral(_ p: CBPeripheral, didOpen channel: CBL2CAPChannel?, error: Error?) {
        MainActor.assumeIsolated {
            guard let cont = openCont else { return }
            openCont = nil
            if let ch = channel {
                cont.resume(returning: ch)
            } else {
                cont.resume(throwing: CableConnError.connectFailed(error?.localizedDescription ?? "openL2CAPChannel failed"))
            }
        }
    }

    func cancel() {
        finishScan(nil)
        if let p = connectingPeripheral { central.cancelPeripheralConnection(p) }
    }
}
